dll/win32/dnsapi/dnsapi/query.c

```diff
@@ -93,6 +93,65 @@ DnsCToW(const CHAR *NarrowString)
     return WideString;
 }
 
+static PCHAR
+DnsWToUTF8(const WCHAR *WideString)
+{
+    PCHAR AnsiString;
+    int AnsiLen = WideCharToMultiByte(CP_UTF8,
+                                      0,
+                                      WideString,
+                                      -1,
+                                      NULL,
+                                      0,
+                                      NULL,
+                                      0);
+    if (AnsiLen == 0)
+        return NULL;
+    AnsiString = RtlAllocateHeap(RtlGetProcessHeap(), 0, AnsiLen);
+    if (AnsiString == NULL)
+    {
+        return NULL;
+    }
+    WideCharToMultiByte(CP_UTF8,
+                        0,
+                        WideString,
+                        -1,
+                        AnsiString,
+                        AnsiLen,
+                        NULL,
+                        0);
+
+    return AnsiString;
+}
+
+static PWCHAR
+DnsUTF8ToW(const CHAR *NarrowString)
+{
+    PWCHAR WideString;
+    int WideLen = MultiByteToWideChar(CP_UTF8,
+                                      0,
+                                      NarrowString,
+                                      -1,
+                                      NULL,
+                                      0);
+    if (WideLen == 0)
+        return NULL;
+    WideLen *= sizeof(WCHAR);
+    WideString = RtlAllocateHeap(RtlGetProcessHeap(), 0, WideLen);
+    if (WideString == NULL)
+    {
+        return NULL;
+    }
+    MultiByteToWideChar(CP_UTF8,
+                        0,
+                        NarrowString,
+                        -1,
+                        WideString,
+                        WideLen);
```

Comment:
You multiplied `WideLen` by `sizeof(WCHAR)` above, but the function expects a character count.

Reply:
You are right MultiByteToWideChar function expects a character count
and RtlAllocateHeap function expect size in bytes.
This code was in base version of function DnsQuery_A.
I pushed updates. Please check them.